### PR TITLE
Drupal 10 readiness

### PR DESCRIPTION
**Pull request by rosiel**
**GitHub Issue**: #63 , supercedes PR #66 

* Other Relevant Links (Google Groups discussion, related pull requests, 
Release pull requests, etc.)

# What does this Pull Request do?

Tells Drupal the jsonld module is compatible with Drupal 10 (~and also Drupal 9~). ~This assertion still needs to be proven, hence this PR is in draft.~  Turned out it is impossible to be D9 and D10 compatible simultaneously, so this now proposes a 3.x branch to accommodate Drupal 10.

# What's new?

* update composer.json
* update info.yml
* update testing scenarios
* (new) Update README.md.

# How should this be tested?

Install the module in Drupal 10 (~and in D9~) and set up an RDF mapping to jsonld and expose it at an endpoint. Does it work? Any messages in the watchdog logs? 

# Additional Notes:
Any additional information that you think would be helpful when reviewing this 
PR.

# Interested parties
Tag (@ mention) interested parties or, if unsure, @Islandora/committers

### CODE REVIEW

File: jsonld.info.yml [hunk 1]
```diff
@@ -2,8 +2,8 @@ name: 'JSON-LD'
...
 package: Web services
-core_version_requirement: ^9
+core_version_requirement: ^9|^10
 dependencies:
-  - hal
-  - serialization
-  - rdf
+  - drupal/serialization
+  - hal/hal
+  - rdf/rdf
```
Comment (adam-vessey) [inline jsonld.info.yml]: Don't think that these should be slash-separated, but colon-separated: `{parent_project}:{project}`  kind of thing... So `serialization`, being a core Drupal module as `drupal:serialization`, but... unsure if we can nicely support the core version disjunction with `hal` and `rdf` having been pulled out from being core modules. As in, should be spec'd here as `drupal:hal` in D9, but then `hal:hal` in D10 (and much the same with `rdf`?)... unsure.

File: .github/workflows/build-2.x.yml [hunk 2]
```diff
@@ -21,8 +21,8 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        php-versions: ["7.4", "8.0", "8.1"]
-        drupal-version: ["9.3.x", "9.4.x", "9.5.x-dev"]
+        php-versions: ["8.0", "8.1"]
+        drupal-version: ["9.5.x-dev", "10.1.x-dev"]
```
Comment (adam-vessey) [inline .github/workflows/build-2.x.yml]:
```suggestion
        drupal-version: ["9.5.x", "10.0.x", "10.1.x-dev"]
```

File: jsonld.info.yml (hunk 1 above)
Comment (rosiel) [reply jsonld.info.yml]: oh man my syntax has been a mess today. Thanks Adam! 

Both RDF and HAL have contrib versions that are compatible with either drupal 9 or 10. So using these even in D9 will help get rid of annoying deprecation messages! However for HAL there is one version (1.x) for D9 and one (2.x) for D10. The composer spec includes both, I hope that works.

**Comment by rosiel** [comment]
Currently stymied by the pdf module, which doesn't have a D10 release though a [patch generated by Rector](https://www.drupal.org/project/pdf/issues/3297841) is RTBC several times over, and folks are clamoring for it to get merged. 

The pdf module is installed in all our workflow testing environments by [islandora_ci](https://github.com/Islandora/islandora_ci/blob/github-actions/travis_setup_drupal.sh#L63-L80).

**Comment by rosiel** [comment]
I've got it running on a Drupal 10 site. @whikloj I'm trying to test but not 100% sure if this is what I'm supposed to do:

I installed restUI and got this error:
`Fatal error: Declaration of Drupal\jsonld\Normalizer\NormalizerBase::supportsNormalization($data, $format = null) must be compatible with Drupal\serialization\Normalizer\NormalizerBase::supportsNormalization($data, ?string $format = null, array $context = []): bool in /var/www/html/web/modules/contrib/jsonld/src/Normalizer/NormalizerBase.php on line 23`

**Comment by whikloj** [comment]
@rosiel that is probably because we don't have return types defined and that (in PHP 8) means we are not correctly `extend`ing the base class.
Based on `10.1.x` version of Drupal, it appears we should have a `bool` return type for that method.
https://git.drupalcode.org/project/drupal/-/blob/10.1.x/core/modules/serialization/src/Normalizer/NormalizerBase.php#L26

You probably want to check the other methods we are overriding to ensure our signatures match the base class.

**Comment by rosiel** [comment]
OK, I updated three files based on running into errors, and now I can load the jsonld version of a node. CI will still fail because of pdf, though.

**Comment by rosiel** [comment]
Installing this on Drupal 9 fails: 

`PHP Fatal error:  Declaration of Drupal\jsonld\Normalizer\NormalizerBase::supportsNormalization($data, ?string $format = null, array $context = []): bool must be compatible with Drupal\serialization\Normalizer\NormalizerBase::supportsNormalization($data, $format = null) in /var/www/html/drupal/web/modules/contrib/jsonld/src/Normalizer/NormalizerBase.php on line 23
`

Does this mean that because the core serializer class' signature has changed, therefore we need a new branch for D10?

**Comment by whikloj** [comment]
Yep
https://git.drupalcode.org/project/drupal/-/blob/9.5.x/core/modules/serialization/src/Normalizer/NormalizerBase.php?ref_type=heads#L34

**Comment by whikloj** [comment]
I guess this will have to be a new branch with a new major version.

**Comment by rosiel** [comment]
I've altered the target branch, changed the workflows (they'll still fail due to pdf), and added a note in the README. For linking purposes, this addresses part of https://github.com/Islandora/documentation/issues/2235.

**Comment by rosiel** [comment]
Since https://github.com/Islandora/islandora_ci/pull/18 was merged, I updated the tests so they don't point to my patch branch of the Islandora CI repo.

**Review by alxp — APPROVED** [review]
Works for me and PHPUnit shows no new warnings.

**Pull request merged**